Multi-user: use forking!

danielbeardsley opened the pull request:
Turns out, webpack uses tons of memory and doesn't seem to release
it after you call watching.close() and null out the reference.

Not exactly sure what's up with that, but process isolation will
save us.

We now run a node process for each user whose webpack dir we want to
watch. We use node's IPC to send messages back and forth between
parent and child.

Noteable things:
* Collection of promises wating for a build has been moved
  into webpack-watcher
* Compiler expiration now kills the whole forked process,
  fully releasing the heap it was using

Closes #4

danielbeardsley committed:
Remove unused function param

This function only takes one param.

danielbeardsley committed:
Compilers: extract colection object

So we have a place to implement expiration.

danielbeardsley committed:
CompilerCollection: use new Map()

It better represents what we're doing here and is
easy to iterate over.

danielbeardsley committed:
squash with initial extract

danielbeardsley committed:
CC: use multiline functions

I'll be adding anotehr line to each of these, might as well
refactor them now.

danielbeardsley committed:
CC: keep track of lastAccessed time

for future expiration logic

danielbeardsley committed:
CC: add expiration

Check for expired compilers every 30s and remove them.

Hopefully this reduces memory usage once it's used in the app.

danielbeardsley committed:
CC: Allow passing in of expire time

This allows the creator of the app to specify the expire time.

danielbeardsley committed:
Example: use the expiration paramter

danielbeardsley committed:
Package lock: bump to our newest version number

danielbeardsley committed:
Webpack Watching: extract a module

We're going to replace this part with a fork() and message
passing.

danielbeardsley committed:
Multi-user: use forking!

Turns out, webpack uses tons of memory and doesn't seem to release
it after you call watching.close() and null out the reference.

Not exactly sure what's up with that, but process isolation will
save us.

We now run a node process for each user who's webpack dir we want to
watch. We use node's IPC to send messages back and forth between
parent and child.

Noteable things:
* Collection of promises wating for a build has been moved
  into webpack-watcher
* Compiiler expiration now kills the whole forked process,
  fully releasing the heap it was using

danielbeardsley committed:
Version bump (feature: forking)

danielbeardsley committed:
Logging: fix undefined var

djmetzle commented on compiler-collection.js:
```diff
@@ -0,0 +1,14 @@
+modules.exports = () => {
+  const compilers = {}
```
This collection might be best represented as a `Map`? Just a thought.

djmetzle commented on multi-user-dev-server.js:
```diff
@@ -24,56 +24,38 @@ const CompilerCollection = require("./compiler-collection.js");
     }
 
     const options = optionsFromUsername(username);
+    options.username = username;
```
Wat? That seems weird. I would think that `optionsFromUsername` would add this value to the `options` object?

danielbeardsley replied:
Yes, but this function is outside the app (i.e. the mapping function is part of the input to the app) so placing an arbitrary restriction on it (that we could easily fulfill ourselves) seemed like a bad idea.

sctice replied:
Fair enough. That might be worth a comment since it looked odd to both of us.

djmetzle commented on multi-user-dev-server.js:
```diff
@@ -24,56 +24,38 @@ const CompilerCollection = require("./compiler-collection.js");
-    // `whenDone` will add pending promises to this list, which will be resolved
-    // when the `watching` handler gets called.
-    let promises = [];
+    const compiler = WebpackWatcher(options);
```
The comment here seems irrelevant now?

danielbeardsley replied:
Removing!

djmetzle commented on webpack-watcher.js:
```diff
@@ -1,2 +1,58 @@
+         child.kill();
+         notifyWatchers()
+      },
+      whenDone
```
Just `whenDone`? Is this a special syntax thing? It seems erroneous.

sctice replied:
It creates a property named `whenDone` with the value of the `whenDone` variable. It's a shorthand for constructing objects from variables.

I don't really see why we're defining it this way instead of in the same way as `close`. Am I missing a use of `whenDone` elsewhere?

danielbeardsley replied:
Moved it inside!

djmetzle commented:
This is really awesome @danielbeardsley!

I do have some critique though, that i hope doesn't come off as too pushy. I think that maybe we should use ES6 classes in a few places. The [watcher object](https://github.com/iFixit/multi-user-dev-server/blob/139ba8a90827de9d833a91981828cb3cf926084f/webpack-watcher.js) seems to be a good candidate for a object proper in it's own right.

Also, i think there is some fork/module/IPC magic going on in the [compiler include file](https://github.com/iFixit/multi-user-dev-server/blob/139ba8a90827de9d833a91981828cb3cf926084f/webpack-compiler.js) that could be separated out into it's own thing. I would like to see a `compiler-forker-entrypoint` module for just kicking off the child pid. I think that separating out the entrypoint could help to understand and follow the logical flow.

Just some pesky structural comments. I hope they're helpful. Feel free to disregard.

CR :medal_military:

sctice-ifixit committed:
[Minor] compiler-collection: Typo

sctice commented on compiler-collection.js:
```diff
@@ -0,0 +1,47 @@
+  const compilers = new Map();
+
+  if (expireAfterSeconds) {
+    setInterval(removeExpiredCompilers, 30 * 1000);
```
It seems to me that this interval should be based on `expireAfterSeconds`. Otherwise, an `expireAfterSeconds < 30` doesn't really do what it advertises.

danielbeardsley replied:
True. But, to be fair, an expiration of 30s or less is probably a bad idea. Either way, I'll make the change.

sctice replied:
> True. But, to be fair, an expiration of 30s or less is probably a bad idea.

This coming from the guy who [made the example 10](https://github.com/iFixit/multi-user-dev-server/pull/6/files#diff-e55bc98107f12615bba2a6649f130228R12) 😁.

sctice commented on compiler-collection.js:
```diff
@@ -0,0 +1,47 @@
+    compiler.lastAccessed = Date.now();
+  }
+
+  const interface = {
```
`interface` is a [future reserved word](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#Future_reserved_keywords). I think it would be a good idea to use something else, maybe `collection`?

danielbeardsley replied:
TIL! Fixing!

sctice commented on webpack-compiler.js:
```diff
@@ -0,0 +1,34 @@
+      case 'isRunning':
+         // If we're not running (building), let them know we're done;
+         // otherwise, they'll find out via the 'built' event.
+         if (!watcher.running) {
```
I think this is assuming a well-behaved parent, who never sends `isRunning` before `watch`.

danielbeardsley replied:
Good call, easy to address.

sctice commented on webpack-watcher.js:
```diff
@@ -0,0 +1,57 @@
+      close: () => {
+         console.log("Killing " + options.username);
+         child.kill();
+         notifyWatchers()
```
It strikes me as odd that this tells the watchers that bundling succeeded.

danielbeardsley replied:
Fair. I'll make it reject!

sctice commented on webpack-watcher.js:
```diff
@@ -0,0 +1,57 @@
+
+   function whenDone() {
+      return new Promise((resolve, reject) => {
+         watchers.push({resolve, reject});
```
I think `watchers` and `notifyWatchers` are misleading names here when we have a thing called `watcher`, and that's _not_ what goes into `watchers`. Maybe something like `whenDoneObservers` or `whenDonePromiseResolvers`?

danielbeardsley replied:
I went for renaming the public interface instead.

sctice commented on webpack-watcher.js:
```diff
@@ -0,0 +1,57 @@
+   console.log("Forking " + options.username);
+   const child = childProcess.fork(__dirname + '/webpack-compiler.js')
+
+   let bundledCallback = () => {}
```
I find this name a little unintuitive because my first reading is to treat "bundled" as an adjective, i.e., a callback that is somehow "bundled". Since it's effectively an event handler, I'd suggest something like `onBundled`.

sctice commented on webpack-watcher.js:
```diff
@@ -0,0 +1,57 @@
+   const child = childProcess.fork(__dirname + '/webpack-compiler.js')
+
+   let bundledCallback = () => {}
+   let watchers = [];
```
This could be a `const` if we emptied the list of watchers instead of replacing it. I think that would be a little more true to the way it's used. Unfortunately, there isn't a super intuitive way to empty an array (unless you think `watchers.length = 0` is intuitive), but if we replaced this with a `Set`, then we could do `watchers.clear()`.

sctice commented:
Cool stuff. I agree with @djmetzle's comments about classes, and I left a few other small suggestions and a couples questions about what may be bugs worth fixing. CR 🎏 but deploy_block ✋ on questions.

danielbeardsley committed:
[minor] remove outdated comment

danielbeardsley committed:
compiler: deal with messages out of order

If the parent forgets to send the watch message, let's
not fail when they send 'isRunning'

danielbeardsley committed:
Watcher: rename callback

Heh, bundledCallback sounds like it was a callback
that is bundled.

danielbeardsley committed:
Watcher: use Set for watcher collection

So we can use 'const' and since the order of these doesn't
really matter.

danielbeardsley committed:
Watcher: rename interface

So it doesn't get confuse with the list of watchers. Also,
move whenDone function into the interface instead of being
a regular function cause it wasn't referenced outside of
that.

danielbeardsley committed:
Watcher: tell watchers the build failed on kill

If there's someone waiting on the build to complete, they should
be told that it failed, not that it was successful.

danielbeardsley committed:
[minor] Comment on an oddity

danielbeardsley committed:
compiler-collection: don't use resereved word

TIL interface is a future reserved word!

djmetzle commented:
CR :fork_and_knife:

danielbeardsley commented:
> Also, i think there is some fork/module/IPC magic going on in the compiler include file that could be separated out into it's own thing

You mean the message sending and receiving? I'm not sure what parts I would pull out of [this file](https://github.com/iFixit/multi-user-dev-server/blob/139ba8a90827de9d833a91981828cb3cf926084f/webpack-compiler.js) to make the code more clear.

I'm confused about this:

>  I would like to see a compiler-forker-entrypoint module for just kicking off the child pid.

But an entrypoint would be the thing you would run in the subprocess, thus it wouldn't do the "kicking off the child pid", that'd be the job of the parent.

danielbeardsley commented:
Thanks for all the suggestions! I liked them all.

un_deploy_block 👍  There's definitely more options for refactoring here, but I don't want to rewrite it entirely, so I'm ok with leaving the code that doesn't need to be touched as-is.

danielbeardsley committed:
Compiler expiration: setInterval sanely

Instead of arbitrarily choosing 30s, let's check for expiration every
(expiretime / 5) so that things can only last about 20% longer than
their expiration time.

danielbeardsley committed:
watcher: fix subtle-ass bug

When an ES6 arrow function is surrounded by brackets, it's treated
as a block and you must explicitly return a value. When it's not,
it's treated as an expression and the value of the expression is
returned.

Returning null here made a further call to Promise.race() resolve
successfully immediately (seems like wacky behavior).

From the Promise.race spec:
  > If the iterable contains one or more non-promise value and/or an
    already resolved/rejected promise, then Promise.race will
    resolve to the first of these values found in the iterable.

So since null is a non-promise value, it resolves immediately to that.

sctice-ifixit committed:
[minor] compiler-collection.js: Typo

sctice commented on compiler-collection.js:
```diff
@@ -0,0 +1,49 @@
+  if (expireAfterSeconds) {
+    // Check for expired things often enough to expire things within
+    // 20% of their true expiration time.
+    setInterval(removeExpiredCompilers, expireAfterSeconds / 5);
```
It occurs to me that a simpler way to do this might be to register a per-compiler timeout whenever we would `updateLastAccessed` for `expireMs` and cancel any existing timeout at the same time. I think it's basically as easy as storing the last accessed time, and then we expire things exactly when we expect to. Does that make sense?

danielbeardsley replied:
Not a bad idea. I think that's probably a bit more intuitive and probably less code. Doing it!

sctice commented on multi-user-dev-server.js:
```diff
@@ -15,61 +16,46 @@ const Express = require("express");
  *                            example/app.js.
+ * @param expireUnusedAfterSeconds Number of seconds after which webpack
+ *                            watchers that are unused will be closed /
+ *                            released. 0 or null == no expiration.
```
I am compelled to comment that this style of indentation is silly, and I don't understand why some people prefer it. Indentation should be consistent so that the amount of whitespace tells us something about nesting rather than depending purely on the size of a name!

Of course, this isn't directed at you. You're just correctly matching style.

sctice commented on webpack-watcher.js:
```diff
@@ -0,0 +1,54 @@
+      whenDone: () => new Promise((resolve, reject) => {
+         watchers.add({resolve, reject});
+         child.send({event: 'isRunning'})
+         console.log("asking isRunning");
```
Something feels weird to me about the `isRunning` message, but I'm having trouble putting my finger on it. It seems somehow superfluous or mismatched to what we're trying to accomplish.

Could we just maintain our own internal state for "asked for build" and "build done" and refer to that in `whenDone` to decide whether to `notifyWatchers()`?

danielbeardsley replied:
The problem is there is no "build started" event that we can listen for to know the state, all we can do is ask "is running". We could poll it fast and let the parent know when it changes, but that's just more code. I'm open to it, but I'm not sure it's a better solution

sctice replied:
Ah, I see. We know when we ask it to start watching, and we know when it finishes building something, but it doesn't tell us when it's started building something. That's a bummer.

sctice commented:
CR 🎏 but deploy_block ✋ on considering a couple more suggestions.

danielbeardsley committed:
Compiler-collection: use individual timeouts for expiration

Much simpler, and more accurate!

danielbeardsley committed:
Logs: remove not-useful logs

These two in particular were added during debugging
and I think they are overly verbose and not that
useful.

danielbeardsley commented:
un_deploy_block 👍  followed your suggestions!

djmetzle commented on compiler-collection.js:
```diff
@@ -2,40 +2,24 @@ module.exports = (expireAfterSeconds) => {
+    if (!compiler || !expireMs) {
+       return;
+    }
+    clearTimeout(compiler.expiredTimeout);
```
Neat.

djmetzle commented:
More CR :gear:

sctice commented:
Sweet! CR 🎏

danielbeardsley committed:
Expiration message: only print when expiring

Not every time we access it, Woops!

djmetzle commented:
CR :stopwatch:

sctice commented:
Whoops! Totally missed that. CR 🎏

danielbeardsley commented:
This has been running fine for a few days. I'm gonna merge it.
